File: src/c/auxlib/dashl.c

/* *** LAST REVISED ON 21-JAN-1994 13:47:51.89 */
/* *** SOURCE FILE: [LONGLIB93.SOURCES.C.AUXLIB]DASHL.C */

#include "my.h"
#include <math.h>

/* Table of constant values */
static real c_b5 = (float)0.;
static real c_b4 = (float).14;
static doublereal c_b9 = .5;
static integer c_n1 = -1;
static integer c__3 = 3;

/* Subroutine */ int dashl_(x, y, n, k, j, l, ix, iy, xmin, dx, ymin, dy)
real *x, *y;
integer *n, *k, *j, *l, *ix, *iy;
real *xmin, *dx, *ymin, *dy;
{
    /* Initialized data */

    static real dl = (float).3;
    static real sl = (float).15;

    /* System generated locals */
    real r__1, r__2;
    doublereal d__1;
    char ch__1[1];

    /* Builtin functions */
    double atan2(), cos(), sin();

    /* External functions */
    extern /* Subroutine */ int plot_(real *, real *, integer *);
    extern /* Subroutine */ int symbol_(real *, real *, real *, char *, real *
	    , integer *, integer *);

    /* Local variables */
    static real a, d, d0;
    static integer i2, i3, i4, i5;
    static real x0, y0, x1, y1, dc, dd;
    static integer la;
    static integer i2y;
    static logical pen;

/* 	MODIFIED FROM LINE D.LONG 24-MAY-1983 AT JPL */

/* 	X	(R) ARRAY OF UNSCALED ABCISSA VALUES */
/* 	Y	(R) ARRAY OF UNSCALED ORDINATE VALUES */
/* 	N	(I) NUMBER OF POINTS IN THE ARRAY */
/* 	K	(I) REPEAT CYCLE OF A MIXED ARRAY (NORMALLY =1) */
/* 	J	(I) >0, SYMBOL AND LINE SYMBOL PLOTTED EVERY JTH POINT */
/* 		    =0, LINE ONLY */
/* 		    <0, SYMBOL ONLY SYMBOL PLOTTED EVERY JTH POINT */
/* 	L	(I) NUMBER OF SYMBOL ,SEE SYMBOL ROUTINE FOR LIST */
/* 	IX,IY	(I) FIRST REVALENT VALUE IN X,Y ARRAY */
/* 	XMIN,YMIN (R) MINIMUM VALUES IN ARRAY (DETERMINED BY SCALE) */
/* 	DX,DY	(R) SCALED (SMOOTHED) INCREMENT (DETERMINED BY SCALE) */

    /* Parameter adjustments */
    --y;
    --x;

    /* Function Body */
/* DASH AND SPACE LENGTH */
    if (*n < 1) {
	goto L110;
    }
    la = *l;
    if (la < 0 || la > 117) {
	la = (float)100.;
    }
    i2 = *ix;
    i2y = *iy;
    i3 = 3;
    if (*j >= 0) {
	i3 = 2;
    }
    if (*dx != (float)0.) {
	x0 = (x[i2] - *xmin) / *dx;
    } else {
	x0 = (float)0.;
    }
    if (*dy != (float)0.) {
	y0 = (y[i2y] - *ymin) / *dy;
    } else {
	y0 = (float)0.;
    }
    plot_(&x0, &y0, &c__3);
    i4 = i3;
    i5 = *n - 1;
    d = (float)0.;
    pen = TRUE_;
    if (*j != 0) {
	if ((i2 - 1) % *j == 0) {
	    ch__1[0] = la;
	    symbol_(&x0, &y0, &c_b4, ch__1, &c_b5, &c_n1, &c_n1);
	}
    }
    if (i5 <= 0) {
	goto L110;
    }
L100:
    if (*dx != (float)0.) {
	x1 = (x[i2] - *xmin) / *dx;
    } else {
	x1 = (float)0.;
    }
    if (*dy != (float)0.) {
	y1 = (y[i2y] - *ymin) / *dy;
    } else {
	y1 = (float)0.;
    }
/* Computing 2nd power */
    r__1 = x1 - x0;
/* Computing 2nd power */
    r__2 = y1 - y0;
    d__1 = (doublereal) (r__1 * r__1 + r__2 * r__2);
    d0 = pow( (double) d__1, (double) c_b9);
    i4 = 3;
    dc = sl;
    if (pen) {
	dc = dl;
    }
    if (d + d0 < dc) {
	if (pen) {
	    i4 = i3;
	}
	plot_(&x1, &y1, &i4);
	if (*j != 0) {
	    if ((i2 - 1) % *j == 0) {
		ch__1[0] = la;
		symbol_(&x1, &y1, &c_b4, ch__1, &c_b5, &c_n1, &c_n1);
	    }
	}
	d = d0 + d;
	x0 = x1;
	y0 = y1;
	goto L105;
    }
    dd = dc - d;
    d = (float)0.;
    a = atan2(y1 - y0, x1 - x0);
    x0 += dd * cos(a);
    y0 += dd * sin(a);
    if (pen) {
	i4 = i3;
    }
    plot_(&x0, &y0, &i4);
    pen = ! pen;
    goto L100;
L105:
    i2 += *k;
    i2y += *k;
    --i5;
    if (i5 > 0) {
	goto L100;
    }
L110:
    return 0;
} /* dashl_ */

File: src/f/auxlib/dashl.f
C
C
	SUBROUTINE DASHL(X,Y,N,K,J,L,IX,IY,XMIN,DX,YMIN,DY)
C
C	MODIFIED FROM LINE D.LONG 24-MAY-1983 AT JPL
C
C	X	(R) ARRAY OF UNSCALED ABCISSA VALUES
C	Y	(R) ARRAY OF UNSCALED ORDINATE VALUES
C	N	(I) NUMBER OF POINTS IN THE ARRAY
C	K	(I) REPEAT CYCLE OF A MIXED ARRAY (NORMALLY =1)
C	J	(I) >0, SYMBOL AND LINE SYMBOL PLOTTED EVERY JTH POINT
C		    =0, LINE ONLY
C		    <0, SYMBOL ONLY SYMBOL PLOTTED EVERY JTH POINT
C	L	(I) NUMBER OF SYMBOL ,SEE SYMBOL ROUTINE FOR LIST
C	IX,IY	(I) FIRST REVALENT VALUE IN X,Y ARRAY
C	XMIN,YMIN (R) MINIMUM VALUES IN ARRAY (DETERMINED BY SCALE)
C	DX,DY	(R) SCALED (SMOOTHED) INCREMENT (DETERMINED BY SCALE)
C
	REAL X(1),Y(1)
	LOGICAL PEN
	DATA DL/.3/,SL/.15/		! DASH AND SPACE LENGTH
	IF(N.LT.1)GOTO 110
	LA=L
	IF (LA.LT.0.OR.LA.GT.117) LA=100.
	I2=IX
	I2Y=IY
	I3=3
	IF(J.GE.0)I3=2
	IF (DX.NE.0.0) THEN
		X0=(X(I2)-XMIN)/DX
	ELSE
		X0=0.0
	ENDIF
	IF (DY.NE.0.0) THEN
		Y0=(Y(I2Y)-YMIN)/DY
	ELSE
		Y0=0.0
	ENDIF
	CALL PLOT(X0,Y0,3)
	I6=I3
	I4=I3
	I5=N-1
	D=0.0
	PEN=.TRUE.
	IF( J.NE.0) THEN
		IF (MOD((I2-1),J).EQ.0) CALL
     $			SYMBOL(X0,Y0,0.14,CHAR(LA),0.0,-1,-1)
	ENDIF
	IF (I5.LE.0) GOTO 110
100	IF (DX.NE.0.0) THEN
		X1=(X(I2)-XMIN)/DX
	ELSE
		X1=0.0
	ENDIF
	IF (DY.NE.0.0) THEN
		Y1=(Y(I2Y)-YMIN)/DY
	ELSE
		Y1=0.0
	ENDIF
	D0=((X1-X0)**2+(Y1-Y0)**2)**.5
	I4=3
	DC=SL
	IF (PEN) DC=DL
	IF (D+D0.LT.DC) THEN
		IF (PEN) I4=I3
		CALL PLOT(X1,Y1,I4)
		IF(J.NE.0)THEN
	IF (MOD((I2-1),J).EQ.0)CALL SYMBOL(X1,Y1,0.14,CHAR(LA),0.0,-1,-1)
		ENDIF
		D=D0+D
		X0=X1
		Y0=Y1
		GOTO 105
	ENDIF
	DD=DC-D
	D=0.0
	A=ATAN2(Y1-Y0,X1-X0)
	X0=X0+DD*COS(A)
	Y0=Y0+DD*SIN(A)
	IF (PEN) I4=I3
	CALL PLOT(X0,Y0,I4)
	PEN=.NOT.PEN
	GOTO 100
105	I2=I2+K
	I2Y=I2Y+K
	I5=I5-1
	IF(I5.GT.0)GOTO 100
110	RETURN
	END
